Add `alternate_email` prop to OIDC userinfo

jmhooper opened the pull request:
**Why**: So that if SPs have users with multiple email addresses, they can access those email addresses

This is a proposal for dealing with:

a.) A feature request from SPs to be able to see all of a user's email addresses
b.) A feature request from SPs to have a mechanism for selecting email addresses when a user signs in with a PIV (i.e. without an email)

Users with multiple email addresses will have the email addresses they did not use to sign in under an `alternate_emails` property. Users with only one email address will have an empty array there:

![image](https://user-images.githubusercontent.com/963654/134726244-309ea049-5223-4fa0-9527-4915313817c6.png)

Additionally, I updated the consent screen to demonstrate that all of the user's email addresses are being shared. We should likely make this change anyway given that we may share a user's alternate email with an SP if they use it to sign in:

![image](https://user-images.githubusercontent.com/963654/134726345-ebc1cc9f-1145-4bba-99b8-a2c9ecf18a69.png)

zachmargolis commented on app/presenters/openid_connect_user_info_presenter.rb:
```diff
@@ -13,6 +13,7 @@ def user_info
       sub: uuid_from_sp_identity(identity),
       iss: root_url,
       email: email_from_sp_identity(identity),
```
This reuses the existing email scope, what if we added a new scope so this attribute becomes opt in?

bonus, as a separate scope it would also trigger the permission screen again right?

aduth commented:
For the consent screen, I might worry that it's easy to miss there being multiple emails shared, given the singular "Email address" label and single line of comma-separated emails. It might be helped with a pluralized "Email addresses" and/or one-email-per-line, though I think it'd be good to get UX weigh-in.

jmhooper commented:
I've added a new scope for the alternate email addresses. I'm thinking through how that should work in the UI for consent.

I'm thinking that we should show all email addresses regardless of which of the scopes is requested. My reasoning here is that any of the email addresses _could_ be shared with the partner even if they aren't in that session. Does that make sense?

I think the other alternative is to split them out into 2 categories. Something like "Email address you sign in with" and "Other email addresses on your account". The "Other email addresses on your account" field would require an empty state.

zachmargolis commented:
> I've added a new scope for the alternate email addresses. I'm thinking through how that should work in the UI for consent. 

I also realized that having a separate scope means this is easier to apply to SAML, it's just another attribute

> I'm thinking that we should show all email addresses regardless of which of the scopes is requested. My reasoning here is that any of the email addresses _could_ be shared with the partner even if they aren't in that session. Does that make sense?
>
> I think the other alternative is to split them out into 2 categories. Something like "Email address you sign in with" and "Other email addresses on your account". The "Other email addresses on your account" field would require an empty state.

I'm imagining this is like a new attribute, so when only the existing email is requested we show that default email (maybe label it as "current email"). Only when additional emails are requested do we show a second line for "additional email addresses" and show the full list there

jmhooper commented:
Okay, I've gone ahead and thrown together a UI that we can start cleaning up. Here is what it like in a few cases. I still need to go through and add some tests for the behavior.

Here is signing in without the alternate_emails scope:
![localhost_3000_sign_up_completed (3)](https://user-images.githubusercontent.com/963654/135651913-05467b9c-f4eb-43eb-9892-004e7d482b37.png)

Here is signing in with the scope, but no other email addresses:
![localhost_3000_sign_up_completed](https://user-images.githubusercontent.com/963654/135651983-48c37e11-9e53-47fd-897b-f19d18add853.png)

Here is what it looks like with a single other email address:
![localhost_3000_sign_up_completed (1)](https://user-images.githubusercontent.com/963654/135652018-279b60fb-6481-46cf-9e88-ccdde623c8c4.png)

Here is what it looks like with more than 2 email addresses:
![localhost_3000_sign_up_completed (2)](https://user-images.githubusercontent.com/963654/135652046-b799b523-34e4-4be3-9963-d8e4c4181600.png)

zachmargolis commented:
> Okay, I've gone ahead and thrown together a UI that we can start cleaning up

yes! love it

zachmargolis commented on app/presenters/openid_connect_user_info_presenter.rb:
```diff
@@ -14,6 +14,8 @@ def user_info
       email: email_from_sp_identity(identity),
       email_verified: true,
+      alternate_emails: alternate_emails_from_sp_identity(identity),
+      alternate_emails_verified: true,
```
I would drop this attribute... the only reason we have `email_verified` is it's part of the spec and we kept it, so since we're inventing a new trait I would not add a corresponding `_verified`

However if we do feel strongly about keeping it, I think it should be an array (that could be zipped with the emails, or maybe even a hash/map? because a single boolean doesn't allow for only some to be verified

zachmargolis commented on app/presenters/openid_connect_user_info_presenter.rb:
```diff
@@ -37,6 +39,10 @@ def email_from_sp_identity(identity)
     EmailContext.new(identity.user).last_sign_in_email_address.email
   end
 
+  def alternate_emails_from_sp_identity(identity)
+    EmailContext.new(identity.user).alternate_email_addresses.map(&:email)
+  end
+
```
we use `EmailContext.new(...)` up above, so let's memoize out a method? save some allocations + queries

```suggestion
  def alternate_emails_from_sp_identity(identity)
    email_context.alternate_email_addresses.map(&:email)
  end

  def email_context
    @email_context ||= EmailContext.new(identity.user)
  end
```

jmhooper commented:
Met up with @anniehirshman-gsa and went over this. Here is the feedback I got:

1. Let's keep all of the email addresses in the same row
1. We should use the label to indicate what is being requested. Just email address if they are requesting the sign in email addresses. If they are requesting all email addresses, change the label to "All email addresses on your account"
1. Each email should be on its own line

And reminder: We still need to consider what happens when information and how consent works there.

jmhooper commented:
@anniehirshman-gsa: Good catch on the typo. I fixed when I added translations

@zachmargolis and @aduth: I added tests, translations, and some other finishing touches. PTAL and lmk what you think!
